### PR TITLE
Flexible offset formatting [1/3]

### PR DESCRIPTION
**Pull request by pitdicker**
There are a lot of issues with our current code to parse time zone offsets. For example `%z`, `%:z`, `%::z` and `%:::z` are all treated identical in the parser. In some cases `Z` is accepted, and there is a hack to accept `UTC`. There is no way to parse seconds.

Formatting is in a better state, but it would be nice if it could be more flexible. For example to be able to print seconds optionally, support padding, or print `Z`.

---

This is part 1 of a series of PRs to improve offset formatting and parsing. (1) adds a flexible formatter, (2) adds a flexible parser and fixes current parsing bugs, and (3) exposes the functionality in the format string.

With this PR it is now possible to specify the format of an offset:
- Is the precision in hours, minutes, seconds, or are minutes and/or seconds optional?
- Are the values seperated by a colon, nothing, or (when parsing) are both acceptable?
- Is `Z` used for `+00:00`?
- Should the hour value be zero-padded, space-padded (before the sign), or not be padded?

### CODE REVIEW

**Review by jtmoon79 — APPROVED** [review]
LGTM

File: src/format/mod.rs [hunk 1]
```diff
@@ -485,6 +485,51 @@ impl Locales {
     }
 }
 
+/// Type for specifying the format of UTC offsets.
```
Comment (djc) [inline src/format/mod.rs:479]: This should be in the same commit where it starts being used. I think I'd prefer it just being called `OffsetFormat` -- the `Utc` prefix seems superfluous.

File: src/format/mod.rs [hunk 2]
```diff
@@ -485,6 +485,51 @@ impl Locales {
...
+    /// Format offset from UTC as hours, and optionally with minutes.
+    /// Any seconds will be rounded to the nearest minute.
+    OptionalMinutes,
+    /// Format offset from UTC as hours and minutes, and optionally seconds.
+    OptionalSeconds,
+    /// Format offset from UTC as hours and optionally minutes and seconds.
+    OptionalMinutesAndSeconds,
+}
+
+/// The seperator between hours and minutes in an offset.
```
Comment (djc) [inline src/format/mod.rs:513]: "seperator" -> "separator" everywhere.

File: src/format/mod.rs [hunk 3]
```diff
@@ -710,144 +710,142 @@ fn format_inner(
...
-                        }
-                        Ok(())
-                    }),
-                    UpperAmPm => time.map(|t| {
-                        result.push_str(if t.hour12().0 {
-                            locale.am_pm[1]
-                        } else {
-                            locale.am_pm[0]
-                        });
+            let ret = match *spec {
```
[the remark this answers is not in the document]
Comment (djc) [reply src/format/mod.rs:713]: So this is old code that was previously not formatted? What changed that rustfmt is now formatting it?

File: src/format/mod.rs [hunk 4]
```diff
@@ -575,7 +575,7 @@ impl UtcOffsetFormat {
     }
 }
 
-#[cfg(any(feature = "alloc", feature = "std"))]
+#[cfg(any(feature = "alloc", feature = "std", test))]
```
[the remark this answers is not in the document]
Comment (pitdicker) [reply src/format/mod.rs:578]: > This doesn't make sense to me. We should instead make the new test(s) conditional on `any(feature = "alloc", feature = "std")`.

Then I am not the only one :smile:. It is a pattern in chrono:
```
$ rg "\#\[cfg\(any\(feature = \"alloc\", feature = \"std\", test\)\)\]"
src/naive/date.rs
6:#[cfg(any(feature = "alloc", feature = "std", test))]
19:#[cfg(any(feature = "alloc", feature = "std", test))]
1249:    #[cfg(any(feature = "alloc", feature = "std", test))]
1293:    #[cfg(any(feature = "alloc", feature = "std", test))]

src/naive/datetime/mod.rs
6:#[cfg(any(feature = "alloc", feature = "std", test))]
15:#[cfg(any(feature = "alloc", feature = "std", test))]
858:    #[cfg(any(feature = "alloc", feature = "std", test))]
902:    #[cfg(any(feature = "alloc", feature = "std", test))]

src/naive/time/mod.rs
6:#[cfg(any(feature = "alloc", feature = "std", test))]
14:#[cfg(any(feature = "alloc", feature = "std", test))]
749:    #[cfg(any(feature = "alloc", feature = "std", test))]
795:    #[cfg(any(feature = "alloc", feature = "std", test))]

src/date.rs
7:#[cfg(any(feature = "alloc", feature = "std", test))]
18:#[cfg(any(feature = "alloc", feature = "std", test))]
336:    #[cfg(any(feature = "alloc", feature = "std", test))]
351:    #[cfg(any(feature = "alloc", feature = "std", test))]

src/datetime/mod.rs
21:#[cfg(any(feature = "alloc", feature = "std", test))]
710:    #[cfg(any(feature = "alloc", feature = "std", test))]
721:    #[cfg(any(feature = "alloc", feature = "std", test))]
755:    #[cfg(any(feature = "alloc", feature = "std", test))]
801:    #[cfg(any(feature = "alloc", feature = "std", test))]
826:    #[cfg(any(feature = "alloc", feature = "std", test))]

src/format/mod.rs
40:#[cfg(any(feature = "alloc", feature = "std", test))]
48:#[cfg(any(feature = "alloc", feature = "std", test))]
50:#[cfg(any(feature = "alloc", feature = "std", test))]
52:#[cfg(any(feature = "alloc", feature = "std", test))]
284:    #[cfg(any(feature = "alloc", feature = "std", test))]
290:    #[cfg(any(feature = "alloc", feature = "std", test))]
422:#[cfg(any(feature = "alloc", feature = "std", test))]
431:#[cfg(any(feature = "alloc", feature = "std", test))]
525:#[cfg(any(feature = "alloc", feature = "std", test))]
578:#[cfg(any(feature = "alloc", feature = "std", test))]
619:#[cfg(any(feature = "alloc", feature = "std", test))]
633:#[cfg(any(feature = "alloc", feature = "std", test))]
646:        #[cfg(any(feature = "alloc", feature = "std", test))]
864:#[cfg(any(feature = "alloc", feature = "std", test))]
882:#[cfg(any(feature = "alloc", feature = "std", test))]
892:#[cfg(any(feature = "alloc", feature = "std", test))]
946:#[cfg(any(feature = "alloc", feature = "std", test))]
976:#[cfg(any(feature = "alloc", feature = "std", test))]
995:#[cfg(any(feature = "alloc", feature = "std", test))]
1064:#[cfg(any(feature = "alloc", feature = "std", test))]

src/format/parse.rs
312:            #[cfg(any(feature = "alloc", feature = "std", test))]
327:            #[cfg(any(feature = "alloc", feature = "std", test))]
```

Shall I open a separate PR to make all the relevant tests conditional?

[the remark this answers is not in the document]
Comment (djc) [reply src/format/mod.rs:578]: Yes, that sounds good.

File: src/format/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pitdicker) [reply src/format/mod.rs:713]: I did some experiments at the time, but couldn't pin it down. In https://github.com/chronotope/chrono/pull/1126#issuecomment-1576229422 I run into the same issue.

[the remark this answers is not in the document]
Comment (djc) [reply src/format/mod.rs:713]: Want to file a rustfmt issue?

File: src/format/mod.rs [hunk 5]
```diff
@@ -485,6 +476,145 @@ impl Locales {
...
+            }
+        };
+        let colons = self.colons == Colon::Colon;
+        write_offsets(result, sign, hours, mins, secs, colons, precision, self.padding)
+    }
+}
+
+#[cfg(any(feature = "alloc", feature = "std"))]
+#[allow(clippy::too_many_arguments)]
+fn write_offsets(
```
Comment (djc) [inline src/format/mod.rs:580]: Given that this is a single-use function with `too_many_arguments`, why not inline it?

File: src/format/mod.rs [hunk 6]
```diff
@@ -660,19 +790,41 @@ fn format_inner(
...
+                        let format = OffsetFormat {
+                            precision: OffsetPrecision::Seconds,
+                            colons: Colon::Colon,
+                            allow_zulu: false,
+                            padding: Pad::Zero,
+                        };
+                        off.map(|&(_, off)| format.format(result, off))
+                    }
+                    TimezoneOffsetTripleColon => {
+                        let format = OffsetFormat {
```
Comment (djc) [inline src/format/mod.rs]: Please inline the bindings here.

File: src/format/mod.rs [hunk 7]
```diff
@@ -485,6 +476,145 @@ impl Locales {
...
+/// The separator between hours and minutes in an offset.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Colon {
+    /// No separator
+    None,
+    /// Colon (`:`) as separator
+    Colon,
+    /// No separator when formatting, colon allowed when parsing.
+    Maybe,
+}
```
Comment (djc) [inline src/format/mod.rs:522]: Style nits: helper types should go below the using code.

File: src/format/mod.rs (hunk 5 above)
Comment (pitdicker) [reply src/format/mod.rs]: Not really. In a previous iteration it was called three times. I have inlined it now.

File: src/format/mod.rs [hunk 8]
```diff
@@ -280,6 +280,63 @@ enum InternalInternal {
...
+    pub colons: Colons,
+    /// Represent `+00:00` as `Z`.
+    pub allow_zulu: bool,
+    /// Pad the hour value to two digits.
+    pub padding: Pad,
+}
+
+impl OffsetFormat {
+    /// Create a new `OffsetFormat`.
+    pub const fn new(
```
Comment (djc) [inline src/format/mod.rs:298]: Given that the `new()` calls will still be spread over the same number of lines in practice, I feel like this method isn't really buying us anything, whereas having the field names in the initializer would make the code a little easier to understand (particularly for the `allow_zulu` field).

File: src/format/mod.rs [hunk 9]
```diff
@@ -280,6 +280,63 @@ enum InternalInternal {
     Nanosecond9NoDot,
 }
 
+/// Type for specifying the format of UTC offsets.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct OffsetFormat {
+    /// See `OffsetPrecision`.
+    pub precision: OffsetPrecision,
+    /// Seperator between hours, minutes and seconds.
```
Comment (djc) [inline src/format/mod.rs:288]: One more instance of `sepe`

**Review by djc — APPROVED** [review]
Okay, please fix up these final nits (unless you disagree) and feel free to merge.

File: src/format/mod.rs (hunk 9 above)
Comment (pitdicker) [reply src/format/mod.rs]: Thank you for being thorough, I thought I had them all.

**Pull request merged**